menu/src/main/java/com/jagrosh/jdautilities/menu/MessagePaginator.java

```diff
@@ -0,0 +1,579 @@
+    }
+
+    // Private method that checks MessageReactionAddEvents
+    private boolean checkReaction(MessageReactionAddEvent event, long messageId)
+    {
+        if(event.getMessageIdLong() != messageId)
+            return false;
+        switch(event.getReactionEmote().getName())
+        {
+            // LEFT, STOP, RIGHT, BIG_LEFT, BIG_RIGHT all fall-through to
+            // return if the User is valid or not. If none trip, this defaults
+            // and returns false.
+            case LEFT:
+            case STOP:
+            case RIGHT:
+                return isValidUser(event.getUser(), event.getGuild());
+            case BIG_LEFT:
+            case BIG_RIGHT:
+                return bulkSkipNumber > 1 && isValidUser(event.getUser(), event.getGuild());
+            default:
+                return false;
+        }
+    }
+
+    // Private method that handles MessageReactionAddEvents
+    private void handleMessageReactionAddAction(MessageReactionAddEvent event, Message message, int pageNum)
+    {
+        int newPageNum = pageNum;
+        switch(event.getReaction().getReactionEmote().getName())
+        {
+            case LEFT:
+                if(newPageNum == 1 && wrapPageEnds)
+                    newPageNum = pages + 1;
+                if(newPageNum > 1)
+                    newPageNum--;
+                break;
+            case RIGHT:
+                if(newPageNum == pages && wrapPageEnds)
+                    newPageNum = 0;
+                if(newPageNum < pages)
+                    newPageNum++;
+                break;
+            case BIG_LEFT:
+                if(newPageNum > 1 || wrapPageEnds)
+                {
+                    for(int i = 1; (newPageNum > 1 || wrapPageEnds) && i < bulkSkipNumber; i++)
+                    {
+                        if(newPageNum == 1 && wrapPageEnds)
+                            newPageNum = pages + 1;
+                        newPageNum--;
+                    }
+                }
+                break;
+            case BIG_RIGHT:
+                if(newPageNum < pages || wrapPageEnds)
+                {
+                    for(int i = 1; (newPageNum < pages || wrapPageEnds) && i < bulkSkipNumber; i++)
+                    {
+                        if(newPageNum == pages && wrapPageEnds)
+                            newPageNum = 0;
+                        newPageNum++;
+                    }
+                }
+                break;
+            case STOP:
+                finalAction.accept(message);
+                return;
+        }
+
+        try {
+            event.getReaction().removeReaction(event.getUser()).queue();
+        } catch(PermissionException ignored) {}
+
+        int n = newPageNum;
+        message.editMessage(renderPage(newPageNum)).queue(m -> pagination(m, n));
+    }
+    
+    private Message renderPage(int pageNum)
+    {
+        return messageFunction.apply(pageNum, pages);
```

Comment:
While I do think the idea is good theoretically, I feel like this should be replaced with an index ordered collection of `Message` instances.
My worry is that by making the provision based on a function, people will opt to generate a new Message instance on each call.
I am not opposed to this, but I'd like @jagrosh's opinion on it before we consider this functionality appropriate to merge.

Reply:
I'm basically generating 1 embed/render currently. How about both?

Reply:
My setup:
```kt
with(MessagePaginator.Builder()) {
    setEventWaiter(Bot.eventWaiter)
    allowTextInput(true)

    setItems(pages.size)
    setLeftRightText("previous", "next")
    setEmbedFunction { p, _ -> pages[p - 1].onHelp(event) }

    build()
}.paginate(event.channel, page)
```

`pages` is a `List<HelpFactory>`, a class that creates MessageEmbeds as required.

Reply:
while I also agree to the Collection idea, atm you can lazy-load embeds as wanted. And let's be honest, it's just a few CPU cycles to generate vs RAM (Imagine 1000 users parsing a Wikipedia page into 20 parts each and all Message and MessageEmbeds loaded, while a lot of users just stop reading in page 4)

Reply:
Theoretically you could externally load them into a collection and poll them yes.
I am only raising the concern for the sake of discussing it, as it is a design decision to make.
~~Also I approve of the Kotlin ❤️~~

Reply:
Keeping the function is fine, but it would be wise to just render them once on init and store them in a list instead of calling the function every time to render